### PR TITLE
github_packages: support --keep-old

### PR DESCRIPTION
**Pull request by Bo98**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

We need this to support any sort of workflow for adding bottles to an existing tag, including `brew dispatch-build-bottle`.

### CODE REVIEW

**Review by MikeMcQuaid — DISMISSED** [review]
Great work 👏🏻

**Comment by dawidd6** [comment]
Are you sure this won't break workflows in linuxbrew-core?

**Comment by Bo98** [comment]
What does `linuxbrew-core` do here? Use `--keep-old` on a non-existent package?

**Comment by dawidd6** [comment]
Yes, I think that's what happens.

**Comment by Bo98** [comment]
I could do this so that it ignores `--keep-old` if a tag doesn't exist?

```diff
diff --git a/Library/Homebrew/github_packages.rb b/Library/Homebrew/github_packages.rb
index ad8839ca2..deb576d26 100644
--- a/Library/Homebrew/github_packages.rb
+++ b/Library/Homebrew/github_packages.rb
@@ -226,21 +226,22 @@ class GitHubPackages
     image_tag = GitHubPackages.image_version_rebuild(version_rebuild)
     image_uri = "#{GitHubPackages.root_url(org, repo, DOCKER_PREFIX)}/#{image_name}:#{image_tag}"
 
-    unless keep_old
-      puts
-      inspect_args = ["inspect", image_uri.to_s]
-      if dry_run
-        puts "#{skopeo} #{inspect_args.join(" ")} --creds=#{user}:$HOMEBREW_GITHUB_PACKAGES_TOKEN"
-      else
-        inspect_args << "--creds=#{user}:#{token}"
-        inspect_result = system_command(skopeo, print_stderr: false, args: inspect_args)
-        if inspect_result.status.success?
-          if warn_on_error
-            opoo "#{image_uri} already exists, skipping upload!"
-            return
-          else
-            odie "#{image_uri} already exists!"
-          end
+    puts
+    inspect_args = ["inspect", image_uri.to_s]
+    if dry_run
+      puts "#{skopeo} #{inspect_args.join(" ")} --creds=#{user}:$HOMEBREW_GITHUB_PACKAGES_TOKEN"
+    else
+      inspect_args << "--creds=#{user}:#{token}"
+      inspect_result = system_command(skopeo, print_stderr: false, args: inspect_args)
+      if keep_old
+        # Tag doesn't exist so ignore --keep-old
+        keep_old = false unless inspect_result.status.success?
+      elsif inspect_result.status.success?
+        if warn_on_error
+          opoo "#{image_uri} already exists, skipping upload!"
+          return
+        else
+          odie "#{image_uri} already exists!"
         end
       end
     end
```

**Comment by dawidd6** [comment]
Guess that will do.

**Comment by MikeMcQuaid** [comment]
@Bo98 that works for me too 👍🏻

**Review by MikeMcQuaid — APPROVED** [review]

File: Library/Homebrew/github_packages.rb [hunk 1]
```diff
@@ -194,7 +201,18 @@ def validate_schema!(schema_uri, json)
     exit 1
   end
 
-  def upload_bottle(user, token, skopeo, formula_full_name, bottle_hash, dry_run:, warn_on_error:)
+  def download(user, token, skopeo, image_uri, root, dry_run:)
+    puts
+    args = ["copy", "--all", image_uri.to_s, "oci:#{root}"]
+    if dry_run
+      puts "#{skopeo} #{args.join(" ")} --src-creds=#{user}:$HOMEBREW_GITHUB_PACKAGES_TOKEN"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/github_packages.rb:208]: Not saying we should but: could moving this skopeo with or without dry run logic into a method (as it's used three times now)

**Review by MikeMcQuaid — APPROVED** [review]

File: Library/Homebrew/github_packages.rb (hunk 1 above)
Comment (Bo98) [reply Library/Homebrew/github_packages.rb:208]: The different args, even ones like `print_stderr` make it a bit tricky.

**Comment by Bo98** [comment]
Added code to `odie` when a duplicate manifest is detected.

File: Library/Homebrew/github_packages.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/github_packages.rb:208]: Fine with me 👍🏻

**Review by MikeMcQuaid — APPROVED** [review]

**Pull request merged**